WebSocket/hub.go

```diff
@@ -0,0 +1,205 @@
+package WebSocket
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"student20_pop/actors"
+	"student20_pop/db"
+	"student20_pop/define"
+	"sync"
+	"time"
+)
+
+const SIG_TRESHOLD = 4
+
+/*
+TODO
+faire des tests
+-Subscribing
+-Unsubscribing
+
+Propagating a message on a channel
+Catching up on past messages on a channel /RAOUl
+
+Publish a message on a channel:
+Update LAO properties ->
+LAO state broadcast ->
+Witness a message ->
+
+Creating a 'event' !
+#check from witness
+#verify if witnessed
+-meeting/ Ouriel
+-roll call/ ouriel
+-discussion(?)
+-poll/ouriel
+-cast vote
+-register attendance
+Meeting state broadcast
+*/
+type hub struct {
+	// the mutex to protect connections
+	connectionsMx sync.RWMutex
+
+	// Registered connections.
+	connections map[*connection]struct{}
+
+	//Response for the sender
+	idOfSender int
+	//msg received from the sender through the websocket
+	receivedMessage chan []byte
+
+	logMx sync.RWMutex
+	log   [][]byte
+
+	organizer *actors.Organizer
+	witness   *actors.Witness
+
+	connIndex int
+}
+
+func NewHub() *hub {
+
+	h := &hub{
+		connectionsMx:   sync.RWMutex{},
+		receivedMessage: make(chan []byte),
+		connections:     make(map[*connection]struct{}),
+		connIndex:       0,
+		idOfSender:      -1,
+		organizer:       actors.NewOrganizer("", "orgDatabase.db"),
+		witness:         actors.NewWitness("", "witDatabase.db"),
+	}
+	//publish subscribe go routine !
+
+	go func() {
+		for {
+			//get msg from connection
+			msg := <-h.receivedMessage
+
+			// check if messages concerns organizer
+			var message []byte = nil
+			var channel []byte = nil
+			var response []byte = nil
+			//handle the message and generate the response, if error, print it in console
+			/*check1, err := h.isForOrganizer(msg)
+			if err != nil {
+				fmt.Print(err)
+			}
+			check2, err2 := h.isForWitness(msg)
+			if err2 != nil {
+				fmt.Print(err2)
+			}
+			if check1 && check2 {
+				fmt.Print("cannot be both witness and organizer")
+			} else if check1 {
+				message, channel, response = h.organizer.HandleWholeMessage(msg, h.idOfSender)
+				fmt.Print(err)
+			} else if check2 {
+				//TODO
+			}*/
+
+			message, channel, response = h.organizer.HandleWholeMessage(msg, h.idOfSender)
+
+			h.connectionsMx.RLock()
+			h.publishOnChannel(message, channel)
+			h.sendResponse(response, h.idOfSender)
+			h.connectionsMx.RUnlock()
```

Comment:
treat the received message as if it was for the organizer, generates a response and sends it through the websocket connection